CONTRIBUTING.md

```diff
@@ -0,0 +1,287 @@
+# Contributing code to Sydent
+
+Everyone is welcome to contribute code to Sydent, provided you are willing to
+license your contributions under the same license as the project itself. In
+this case, the [Apache Software License v2](LICENSE).
+
+### Create a virtualenv
+
+To contribute to Sydent, ensure you have Python 3.7 or newer and then run:
+
+```bash
+python3 -m venv venv
+./venv/bin/pip install -e '.[dev]'
+```
+
+This creates an isolated virtual Python environment ("virtualenv") just for
+use with Sydent, then installs Sydent along with its dependencies, and lastly
+installs a handful of useful tools
+
+If you get `ConnectTimeoutError`, this is caused by slow internet whereby 
+`pip` has a default time out of _15 sec_. You can specify a larger timeout 
+by passing `--timeout 120` to the `pip install` command above.
+
+Finally, activate the virtualenv by running:
+
+```bash
+source ./venv/bin/activate
+```
+
+Be sure to do this _every time_ you open a new terminal window for working on
+Sydent. Activating the venv ensures that any Python commands you run (`pip`,
+`python`, etc.) use the versions inside your venv, and not your system Python.
+
+When you're done, you can close your terminal or run `deactivate` to disable
+the virtualenv.
+
+<!-- TODO
+### Run the tests
+
+To make sure everything is working as expected, run the unit tests:
+
+```bash
+tox -e py
```

Comment:
I think changing this will let you uncomment this section:

```suggestion
python -m twisted.trial tests
```

Reply:
Thanks, that sounds good.

Thanks to @H-Shay I noticed that the README includes instructions, but says to use `trial tests`.

I seem to have forgotten the `-e` flag to `pip install -e .[dev]` and notice that:

- `python -m twisted.trial tests` works
- but `trial tests` fails

(I only just realised that I probably forgot `-e` and that might be the cause.)

I never really appreciated the difference there, but I guess `python -m` treats all subdirs of the current working dir as modules, whereas `trial tests` expects the `tests` modules to be installed.

I guess I will leave a hint about that, since this tripped me up for a few minutes and I suspect it could be confusing to someone not used to Python's modules.

Reply:
How is `trial tests` failing? Usually this is a path issue (it is finding the wrong `trial` binary). Using `python -m twisted.trial` should generally be the same... 🤷

Reply:
At first, it failed saying that it can't find `tests`.

Then I did `pip install .[dev]` and it now finds the tests but can't find the templates.

Only with `pip install -e .[dev]` does `trial` work.

I think this is because invoking `trial` will only use the modules available in your venv, but `python -m twisted.trial` seems to have the current directory on Python's path? I'm not ultimately that sure either.